[MXNET-854] SVRG Optimization in Python Module API

StephanieYuan opened the pull request:
## Description ##
SVRG stands for Stochastic Variance Reduced Gradient, which is an optimization technique that complements SGD. It is introduced in the paper _[Accelerating Stochastic Gradient Descent using Predicative Variance Reduction](https://papers.nips.cc/paper/4937-accelerating-stochastic-gradient-descent-using-predictive-variance-reduction.pdf)_ in 2013. 
Key Characteristics of SVRG: 
- Explicit Variance Reduction
- Converge much faster compared to SGD on strongly convex smooth functions   

SVRG Module should:
- seamlessly supports both dense and sparse data, run on CPU and GPU instances on single machine and in distributed setting.
- minimize the API differences with Python Module API.

## Checklist ##
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-854], where MXNET-854 refers to the relevant [JIRA issue](https://issues.apache.org/jira/browse/MXNET-854) created (except PRs with tiny changes)
- [x] All changes have test coverage:
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- SVRGModule inherits from Python Module and minimizes the API changes of Module API. Both intermediate level and high level APIs are supported
- Functional tests and unittests have been tested for SVRGModule and SVRG Optimizer.

## Comments ##
- Design decisions of why using Python Module instead of Optimizer can be found in the design doc [here](https://cwiki.apache.org/confluence/display/MXNET/SVRG+Optimization+in+MXNet+Python+Module)
- This version of SVRGModule supports single machine run with single cpu, single gpu and multiple-gpus. Distributed SVRGModule is soon to be updated.
- Benchmarking has been performed on this single machine version SVRGModule on single cpu on linear regression model. Distributed SVRGModule to be updated soon. Benchmarking results can be found in contrib/svrg_optimization_python/benchmarks.

sandeep-krishnamurthy commented:
Thanks @StephanieYuan - Welcome to MXNet community!

@piiswrong @eric-haibin-lin @anirudhacharya @Roshrini @vandanavk - You will be interested to have a look at this.

sandeep-krishnamurthy commented on contrib/svrg_optimization_python/test_svrg_train.py:
```diff
@@ -0,0 +1,84 @@
+            mod.forward_backward(data_batch=batch)
+            mod.update()
+            mod.update_metric(metrics, batch.label)
+        print('Epoch[%d] Time cost=%.3f', e, metrics.get())
```
Please use logging.

sandeep-krishnamurthy commented on contrib/svrg_optimization_python/test_svrg_train.py:
```diff
@@ -0,0 +1,84 @@
+    :return: an instance of mx.io.NDArrayIter
+    :return: an instance of mx.mod.svrgmodule for performing SVRG optimization
+    """
+    mx.random.seed(42)
```
1. Please do not use fixed seed. You can see other tests under tests/
2. Also, what are we asserting here?

sandeep-krishnamurthy commented on contrib/svrg_optimization_python/test_svrg_train.py:
```diff
@@ -0,0 +1,84 @@
+    return di, mod
+
+# run as a script
+if __name__ == "__main__":
```
should it be? Please see other tests.

```
if __name__ == '__main__':
    import nose
    nose.runmodule()
```

StephanieYuan commented:
Move svrg_optimization to python/contrib package.

vandanavk commented on tests/python/unittest/test_contrib_svrg_optimizer.py:
```diff
@@ -0,0 +1,101 @@
+
+
+def test_init_svrg_optimizer():
+    di, mod = create_network()
```
'di' is unused. Can be changed to '_'

vandanavk commented on tests/python/unittest/test_contrib_svrg_optimizer.py:
```diff
@@ -0,0 +1,101 @@
+
+
+def test_svrg_optimizer_constructor():
+    _, mod = create_network()
```
both _ and mod are not used here

vandanavk commented on example/svrg_module/common.py:
```diff
@@ -0,0 +1,78 @@
+
+def calc_expectation(grad_dict, count):
+    for key in grad_dict.keys():
+        grad_dict[str.format(key+"_expectation")] = mx.ndarray.sum(grad_dict[key], axis=0)/count
```
Is there a chance for Divide By Zero?

StephanieYuan replied:
count refers to number of batches which by default is one, as defined in the NDArrayIter, renamed to num_batch in the latest revision for clarity.

vandanavk commented on example/svrg_module/common.py:
```diff
@@ -0,0 +1,78 @@
+    for i in range(len(param_names)):
+        diff_sqr = mx.ndarray.square(mx.nd.subtract(grad_dict[param_names[i]],
+                                                    grad_dict[str.format(param_names[i]+"_expectation")]))
+        grad_dict[str.format(param_names[i] + "_variance")] = mx.ndarray.sum(diff_sqr, axis=0) / count
```
Divide By Zero here?

vandanavk commented on example/svrg_module/data_reader.py:
```diff
@@ -0,0 +1,44 @@
+
+    from sklearn.datasets import load_svmlight_file
+
+    feature_dim = 90
```
Please add a comment referring to the source for this number 90.

vandanavk commented on example/svrg_module/data_reader.py:
```diff
@@ -0,0 +1,44 @@
+    feature_stds = np.sqrt(np.square(train_features - feature_means).mean(axis=0))
+
+    train_features = (train_features - feature_means) / feature_stds
+    train_labels = (train_labels - label_mean) / label_std
```
Any chance of Divide By Zero here?

StephanieYuan replied:
These are standard deviations calculated w.r.t to real dataset. As a sanity check, these values are not zeros.

vandanavk commented on example/svrg_module/example_api_train.py:
```diff
@@ -0,0 +1,86 @@
+
+# run as a script
+if __name__ == "__main__":
+    num_epoch = 100
```
Can this be a user-defined param?

vandanavk commented on example/svrg_module/example_inference.py:
```diff
@@ -0,0 +1,89 @@
+            mod.update_metric(metrics, batch.label)
+
+    y = mod.predict(val_iter)
+    assert y.shape == (200, 1)
```
Add a comment about why the comparison is with (200,1)

vandanavk commented on example/svrg_module/example_inference.py:
```diff
@@ -0,0 +1,89 @@
+
+# run as a script
+if __name__ == "__main__":
+    num_epoch = 100
```
user-defined param possible?

vandanavk commented on example/svrg_module/train.py:
```diff
@@ -0,0 +1,44 @@
+from data_reader import read_year_prediction_data
+
+parser = argparse.ArgumentParser()
+parser.add_argument('-e', dest='epochs', help='number of epochs for training phase', type=int, required=True)
```
what is the default value?

vandanavk commented on example/svrg_module/train.py:
```diff
@@ -0,0 +1,44 @@
+parser.add_argument('-f', dest="updateFreq", help="update frequency for SVRGModule", type=int, default=2, required=True)
+parser.add_argument('-b', dest="batch_size", help="define the batch size for training", type=int,
+                    default=100, required=False)
+parser.add_argument('-m', dest='metrics', help="create eval metric", type=str, required=False)
```
Any default value?

vandanavk commented on example/svrg_module/train.py:
```diff
@@ -0,0 +1,44 @@
+parser.add_argument('-b', dest="batch_size", help="define the batch size for training", type=int,
+                    default=100, required=False)
+parser.add_argument('-m', dest='metrics', help="create eval metric", type=str, required=False)
+parser.add_argument('--gpus', type=str, help='list of gpus to run, e.g. 0 or 0,2,5. empty means using cpu')
```
Default value?

StephanieYuan replied:
Default is mx.cpu() as defined in line 35.

vandanavk commented on tests/python/unittest/test_contrib_svrg_module.py:
```diff
@@ -0,0 +1,86 @@
+        model = mx.symbol.LinearRegressionOutput(data=model, label=y, name="out")
+        return model
+
+    n, m = 128, 100
```
Please add a comment explaining the values 128, 100

vandanavk commented on tests/python/unittest/test_contrib_svrg_optimizer.py:
```diff
@@ -0,0 +1,101 @@
+    weights = np.array([1.0, 2.0])
+    train_label = train_data.dot(weights)
+
+    batch_size = 32
```
Can this be user-defined?

anirudhacharya replied:
this is a unit test. how will this be user defined?

StephanieYuan commented on python/mxnet/contrib/svrg_optimization/svrg_optimizer.py:
```diff
@@ -0,0 +1,133 @@
+
+        name = self._check_index(index)
+
+        if "full".lower() in name:
```
Resolved.

StephanieYuan commented:
@xcgoner Thank you for the pointer! I'd like to try out the improvements in the later iterations of this project.

anirudhacharya commented on tests/python/unittest/test_contrib_svrg_module.py:
```diff
@@ -0,0 +1,87 @@
+import numpy as np
+
+
+def set_up():
```
SVRGModule lacks test coverage. Can you add tests for `save_checkpoint`, `load`, `module.forward`, `module.predict`.

anirudhacharya commented on python/mxnet/contrib/svrg_optimization/README.md:
```diff
@@ -0,0 +1,41 @@
+## SVRG Optimization in Python Module
+
+### Problem Description
+SVRG stands for Stochastic Variance Reduced Gradient, which was first introduced in the paper _Accelerating Stochastic 
```
This file should be removed and SVRG description and characteristics should go into the documentation of the SVRG optimizer.

anirudhacharya commented on python/mxnet/contrib/svrg_optimization/README.md:
```diff
@@ -0,0 +1,41 @@
+* test_svrg_module.py: unittests for SVRGModule API
+* test_svrg_optimizer.py: unittests for SVRGOptimizer API
+
+### Examples:
```
add these in the README of the examples folder.

anirudhacharya commented on python/mxnet/contrib/svrg_optimization/README.md:
```diff
@@ -0,0 +1,41 @@
+* examples/svrg_module/example_api_train.py: a demo script for training a SVRGModule using intermediate level api and high level api
+* examples/svrg_module/example_inference.py: a demo script for SVRGModule inference
+
+#### Benchmarks:
```
as the benchmark graphs are present in the "examples" folder, move these lines to the README in the examples folder.

anirudhacharya commented on python/mxnet/contrib/svrg_optimization/README.md:
```diff
@@ -0,0 +1,41 @@
+
+#### Testing:
+
+Unit Tests:
```
remove these lines about the Unit Tests.

anirudhacharya commented on example/svrg_module/linear_regression/README.md:
```diff
@@ -0,0 +1,7 @@
+data_reader.py to download the data. 
+
+This example trains a linear regression model using SVRGModule. Logs of the training results can be found in 
+experiments.log
```
where is experiments.log?

StephanieYuan replied:
It will be generated when runs the train script.

aaronmarkham commented on example/svrg_module/README.md:
```diff
@@ -0,0 +1,26 @@
+## SVRG Module Example
+
```
Introduce what it is as well. From your other document:
"SVRGModule is an extension to the Module API that implements SVRG (Stochastic Variance Reduced Gradients) optimization ..."

aaronmarkham commented on example/svrg_module/README.md:
```diff
@@ -0,0 +1,26 @@
+## SVRG Module Example
+
+#### API Usage Example
+SVRGModule provides both high-level and intermediate-level APIs while minimizing the changes with Module API.  
```
Link back to the docs. This should link back to this spot - where you should add an entry for this module.
https://github.com/apache/incubator-mxnet/blob/master/docs/api/python/index.md#contrib-package
Since this comes from the example folder and outside of the docs, use a full static link.

anirudhacharya commented on tests/python/unittest/save_000800.json:
```diff
@@ -1,188 +0,0 @@
-{
```
Why delete this file and `legacy_ndarray.v0`?

anirudhacharya commented:
Also the way `svrg_optimizer.py` is written it leaves a few questions open. 

From your design document I understand that SVRGOptimizer cannot be used with Module API which is why you wrote the SVRGModule API which will use SVRGOptimizer under the hood instead of the usual SGD or other optimizers. Can you explain why in greater detail?

Also if this is indeed the case, then the **SVRGOptimizer class needs to be declared private using a leading underscore in its name**, else it will be very misleading to the user who might try to use SVRGOptimizer with the plain vanilla Module API. Also it should be made clear in the doc string for the SRVGOptimizer that this optimizer is to be used with the SVRGModule alone.

Also you might want to consider moving the SVRGOptimizer to the `optimizer.py` file and keep SVRGModule alone in the `contrib` folder. But I am not sure of this, you might want to get a second opinion on this. @piiswrong @eric-haibin-lin

sandeep-krishnamurthy commented:
Great progress here. Thanks everyone. 
@StephanieYuan 
1. Please rebase and resolve conflicts. 
2. Please comment about tests added and coverage.
3. We had discussed in above comment, to verify this implementation with MNIST to confirm correctness per the claims in paper. Please comment.

@anirudhacharya - I would still prefer keeping SVRG optimizer in contrib and allow it to evolve. As mentioned in above comments and design doc, Other combination like Adam+SVRG could also be used but not benchmarked. I would set the default optimizer for SVRGModule to be SVRGOptimizer but not remove the option altogether and make SVRGOptimizer internal. Thoughts?

StephanieYuan commented:
Thanks everyone for the feedback and input! 

I've rebased and resolved the conflict with master.  I have also added some more unittests for SVRGModule especially on verifying SVRG gradients calculations and the test coverage is now:
mxnet/contrib/svrg_optimization/svrg_module.py                        175     21    88%
mxnet/contrib/svrg_optimization/svrg_optimizer.py                      38      1    97%

@sandeep-krishnamurthy I'm working on creating the MNIST example but I'm expecting it will take a few more days as I have other tasks in queue. Maybe I can include the example in my next PR on SVRG in distributed mode?

@anirudhacharya SVRGOptimizer is designed to be used with SVRGModule alone. The actual SVRG optimization logic is implemented in the SVRGModule and the SVRGOptimizer is designed to assist the full gradients accumulation in the KVStore. I've made SVRGOptimizer a private class and added the reasonings behind the design in greater detail in the docs/api/python/contrib/svrg_optimization. Hopefully it clarifies things a bit. 

@aaronmarkham I've added pointer about the svrg_optimization in index.md and refined the docs in the example/svrg_module/README.

sandeep-krishnamurthy commented on tests/python/unittest/test_contrib_svrg_module.py:
```diff
@@ -0,0 +1,295 @@
+    mod.init_params()
+    mod.init_optimizer(optimizer='sgd', optimizer_params={'learning_rate': 0.1})
+    mod.update()
+    mod.save_checkpoint('test', 0, save_optimizer_states=True)
```
Recommended to use tempfile. This avoid race condition in file access from other tests and also clean up is easy. Something like:
```
import tempfile
import os
....
 tmp = tempfile.mkdtemp()
 tmpfile = os.path.join(tmp, 'svrg_test')
mod.save_checkpoint(tmpfile, .....)
```

sandeep-krishnamurthy commented on tests/python/unittest/test_module.py:
```diff
@@ -815,4 +815,4 @@ def test_forward_types():
 if __name__ == '__main__':
     import nose
-    nose.runmodule()
+    nose.runmodule()
```
nit: add a new line (pep)

anirudhacharya commented:
@sandeep-krishnamurthy it now looks better after SVRGOptimizer was made private. Including SVRGOptimizer in `optimizer.py` file instead of the `contrib` folder was just a thought. I am okay either ways. Will take a look at @StephanieYuan 's changes and approve it accordingly.

StephanieYuan commented:
@anirudhacharya I'll fix it soon. Is there a way that I can see the rendered view before it gets pushed?

anirudhacharya commented:
@StephanieYuan some of the issues are with the content of the docs and few are with how the content renders. @aaronmarkham might be able to help with how the docs renders. You will probably have to build the docs and have to deploy it locally.

aaronmarkham commented:
I just updated the how-to guide for building the docs.
https://cwiki.apache.org/confluence/display/MXNET/How+to+Build+the+Website
Let me know if you have any questions.

StephanieYuan commented:
@aaronmarkham Thanks for the pointer! 
I'm thinking to exclude SVRGModule from module.md since SVRGModule will be a contrib package at current stage.

anirudhacharya commented on python/mxnet/contrib/svrg_optimization/svrg_module.py:
```diff
@@ -0,0 +1,590 @@
+        # update idx2name to include new keys
+        for key in self._param_dict[0].keys():
+            max_key = max(list(idx2name.keys()))
+            max_key += 1
```
you can increment by 1 in line 187 itself

anirudhacharya commented on python/mxnet/contrib/svrg_optimization/svrg_module.py:
```diff
@@ -0,0 +1,590 @@
+            self.update_freq = update_freq
+        else:
+            raise TypeError("update_freq in SVRGModule must be a positive integer to represent the frequency for "
+                            "calculating full gradients")
```
then you should probably have appropriately different messages for the two errors instead of a generic message for both types of errors.

StephanieYuan commented:
@anirudhacharya Sounds good! It's a Jupyter Notebook. I'll add it to the example folder.
